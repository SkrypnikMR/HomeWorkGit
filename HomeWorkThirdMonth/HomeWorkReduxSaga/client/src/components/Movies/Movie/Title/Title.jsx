import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Button from '../../../Button';
import Input from '../../../Input';
import './Title.scss';
import { NotificationManager } from 'react-notifications';


const Title = ({ name, id, title, updateMovie, deleteMovie }) => {
    const [state, setState] = useState({
        isWaitForChange: false,
        changeParam: 'title',
        newData: '',
    });
    const handleOnChange = (e) => setState({ ...state, newData: e.target.value });
    const handleDeleteItem = () => deleteMovie(id);
    const handleOnDoubleClick = () => setState((state) => ({ ...state, isWaitForChange: !state.isWaitForChange }));
    const handleOnBlur = async (e) => {
        if (e.target.value === '') {
            setState({ ...state, isWaitForChange: false });
            return NotificationManager.error('Please Enter newData for field', 'INPUT ERROR', 2000);
        }
        const { changeParam, newData } = state;
        setState({ ...state, isWaitForChange: false });
        await updateMovie({ changeParam, newData, id: id });
    }

    return (
        <div className="movie__header" >
            {
                state.isWaitForChange ? (
                    <Input
                        name={name}
                        onBlurInput={handleOnBlur}
                        onChangeInput={handleOnChange}
                        placeholder='enter new title'
                        value={state.newData}
                    />
                )
                    : (
                        <h5 name={name} onDoubleClick={handleOnDoubleClick}>
                            {title}
                        </h5>
                    )
            }
            <Button onButtonClick={handleDeleteItem} title="X" />
        </div>
    );
}

Title.propTypes = {
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    updateMovie: PropTypes.func.isRequired,
    title: PropTypes.string
}

export default Title;