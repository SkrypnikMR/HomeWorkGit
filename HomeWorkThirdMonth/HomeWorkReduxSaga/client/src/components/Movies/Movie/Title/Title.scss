.movie__header{
        border-radius: 10px 10px 0 0;
        position: relative;
        top:-10px;
        background-color: #f5da0b;
        width: 100%;
        z-index: 100;
        padding: 0;
        display: grid;
        justify-content: center;
        align-items: center;
        font-weight: bold;
        input {
            height: 40px;
            width: 90%;
            margin: 15px;
            z-index: 1000;
            position: relative;
            overflow-x: hidden;
            background-color: #f5da0b;
            border: none;
            color: white;
            &::placeholder{
                color:white;
                font-weight: bold;
            }
        }
        button {
              font-weight: bold;
            position: absolute;
            top:0px;
            right: 10px;
            padding-right: 10px;
            width: 50px;
            transition-duration: .5s;
            &:hover{
                font-weight: bold;
                transition-duration: .5s;
                font-size: 20px;
                right:15px;
                color: #1B264F;
    font-family: 'Play', sans-serif;
            }
        }
    }