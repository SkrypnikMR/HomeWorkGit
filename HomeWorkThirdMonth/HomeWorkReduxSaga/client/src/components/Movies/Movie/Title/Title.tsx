import React, { useState } from 'react';
import Button from '../../../Button';
import Input from '../../../Input';
import './Title.scss';
import { NotificationManager } from 'react-notifications';

interface ITitle {
    name: string;
    id: string;
    title: string;
    updateMovie: (id: IpdateMovie) => void;
    deleteMovie: (id: string) => void;
}

interface IpdateMovie {
    id: string;
    changeParam: string;
    newData: string;
}

const Title = ({ name, id, title, updateMovie, deleteMovie }: ITitle) => {
    const [state, setState] = useState({
        isWaitForChange: false,
        changeParam: 'title',
        newData: '',
    });
    const handleOnChange = (e :React.ChangeEvent<HTMLInputElement>) => setState({ ...state, newData: e.target.value });
    const handleDeleteItem = () => deleteMovie(id);
    const handleOnDoubleClick = () => setState((state) => ({ ...state, isWaitForChange: !state.isWaitForChange }));
    const handleOnBlur = async (e :React.ChangeEvent<HTMLInputElement>) => {
        if (e.target.value === '') {
            setState({ ...state, isWaitForChange: false });
            return NotificationManager.error('Please Enter newData for field', 'INPUT ERROR', 2000);
        }
        const { changeParam, newData } = state;
        setState({ ...state, isWaitForChange: false });
        await updateMovie({ changeParam, newData, id: id });
    }

    return (
        <div className="movie__header" >
            {
                state.isWaitForChange ? (
                    <Input
                        name={name}
                        onBlurInput={handleOnBlur}
                        onChangeInput={handleOnChange}
                        placeholder='enter new title'
                        value={state.newData}
                    />
                )
                    : (
                        <h5 onDoubleClick={handleOnDoubleClick}>
                            {title}
                        </h5>
                    )
            }
            <Button onButtonClick={handleDeleteItem} title="X" />
        </div>
    );
}



export default Title;