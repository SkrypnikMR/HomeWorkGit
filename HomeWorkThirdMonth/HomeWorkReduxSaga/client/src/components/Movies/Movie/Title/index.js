import Title from './Title.jsx';
import { connect } from 'react-redux';
import { deleteMovie, updateMovie } from '../../../../store/movies/actions';
import { getMovieField } from '../../../../store/movies/selectors';

const mapStateToProps = (state, props) => ({
    title: getMovieField(state, props),
});

const mapDispatchToProps = { deleteMovie, updateMovie };

export default connect(mapStateToProps, mapDispatchToProps)(Title);