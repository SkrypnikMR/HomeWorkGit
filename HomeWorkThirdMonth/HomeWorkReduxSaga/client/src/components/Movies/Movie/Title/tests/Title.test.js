import React from 'react';
import { NotificationManager } from 'react-notifications';
import { shallow, mount } from 'enzyme';
import Title from '../title';
import { act } from 'react-dom/test-utils';


jest.mock('react-notifications', () => ({
    NotificationManager: {
        error: jest.fn()
    }
}))

describe('Title', () => {
    let props;
    it('Should match snapshot', () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const component = shallow(<Title {...props} />);
        expect(component.html()).toMatchSnapshot();
    });
    it('Should render Input ', () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const component = mount(<Title {...props} />);
        act(() => component.find('h5').props().onDoubleClick());
        component.update();
        expect(component.find('h5')).toHaveLength(0);
        expect(component.find('Input')).toHaveLength(1);
    });
    it('Should call deleteMovie', () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const component = mount(<Title {...props} />);
        act(() => component.find('Button').props().onButtonClick());
        component.update();
        expect(props.deleteMovie).toHaveBeenCalledWith(props.id);
    });
    it('Should change Input value ', () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const testValue = 'testValue';
        const component = mount(<Title {...props} />);
        act(() => component.find('h5').props().onDoubleClick());
        component.update();
        act(() => component.find('Input').props().onChangeInput({ target: { value: testValue } }));
        component.update();
        expect(component.find('Input').props().value).toBe(testValue);
    });
    it('Should blur Input but with empty value ', async () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const testValue = '';
        const component = mount(<Title {...props} />);
        act(() => component.find('h5').props().onDoubleClick());
        component.update();
        await act(async () => await component.find('Input').props().onBlurInput({ target: { value: testValue } }));
        component.update();
        expect(NotificationManager.error).toHaveBeenCalled();
    });
    it('Should blur Input without empty value ', async () => {
        props = {
            name: 'title',
            id: 1,
            title: 'someTitle',
            updateMovie: jest.fn(),
            deleteMovie: jest.fn(),
        }
        const testValue = 'someValue';
        const component = mount(<Title {...props} />);
        act(() => component.find('h5').props().onDoubleClick());
        component.update();
        act(() => component.find('Input').props().onChangeInput({ target: { value: testValue } }));
        component.update();
        await act(async () => await component.find('Input').props().onBlurInput({ target: { value: testValue } }));
        component.update();
        expect(props.updateMovie).toHaveBeenCalledWith({ changeParam: props.name, id: props.id, newData: testValue })
    });
});
