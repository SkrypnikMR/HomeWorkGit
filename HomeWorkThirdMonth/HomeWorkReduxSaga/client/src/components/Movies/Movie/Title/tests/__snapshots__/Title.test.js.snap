// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title Should match snapshot 1`] = `"<div class=\\"movie__header\\"><h5 name=\\"title\\">someTitle</h5><button class=\\"custom__button\\">X</button></div>"`;
